packages/vscode-extension/src/webview/components/UrlBar.tsx

```diff
@@ -35,16 +35,50 @@ function ReloadButton({ project, disabled }: ReloadButtonProps) {
   );
 }
 
-function UrlBar({ project, disabled }: UrlBarProps) {
+function UrlBar({ project, disabled, resetUrlHistory }: UrlBarProps) {
+  const MAX_URL_HISTORY_SIZE = 20;
+  const MAX_RECENT_URL_SIZE = 5;
   const [urlList, setUrlList] = useState<{ name: string; id: string }[]>([]);
+  const [recentUrlList, setRecentUrlList] = useState<{ name: string; id: string }[]>([]);
+  const [urlHistory, setUrlHistory] = useState<string[]>([]);
+
   useEffect(() => {
     function handleNavigationChanged(navigationData: { displayName: string; id: string }) {
-      const newRecord = { name: navigationData.displayName, id: navigationData.id };
-      setUrlList((urlList) => [
-        newRecord,
-        ...urlList.filter((record) => record.id !== newRecord.id),
-      ]);
+      if (navigationData.id === "{}") {
+        return;
+      }
+
+      function removeDynamicPostfix(url: string): string {
+        return url.split("?")[0];
+      }
+
+      const newRecord = {
+        name: navigationData.displayName, //removeDynamicPostfix(), TODO
+        id: navigationData.id,
+      };
+      const isNotInHistory = !urlHistory.length || urlHistory[0] !== newRecord.id;
+
+      setUrlList((urls) => [newRecord, ...urls.filter((record) => record.id !== newRecord.id)]);
```

Comment:
This could better show intent if you did:
```
function moveAsMostRecent(urls, url) {
  return [newUrl, ...urls.filter((record) => record.id !== newRecord.id)]
}
```

then you can replace this call and call to `setRecentUrlList` below.